src/drivers/driver_wired_sonic.c

```diff
@@ -0,0 +1,405 @@
+/*
+ * Wired Ethernet driver interface for Sonic
+ * Copyright (c) 2005-2009, 
+ * Copyright (c) 2004
+ *
+ * This software may be distributed under the terms of the BSD license.
+ * See README for more details.
+ */
+
+#include <unistd.h>
+#include <netdb.h>
+#include <netinet/in.h>
+#include "includes.h"
+#include "common.h"
+#include "driver.h"
+#include "driver_wired_common.h"
+#include "ap/sta_info.h"
+#include "radius/radius_attr_parse.h"
+#include "common/eapol_common.h"
+
+#define STATUS_COPY(_status)  static void _status##_##copy(char *intf, clientStatusReply_t *reply, u8 *addr, void *param)
+#define STATUS_ENTER(_status, _intf, _reply, _addr, _param)  _status##_##copy(_intf, _reply, _addr,  _param)
+
+
+#define STATUS_MALLOC(__status, _param, _reply)  \
+  do { \
+    unsigned int _len = 0; \
+    if (AUTH_SUCCESS == __status) \
+    { \
+      struct sta_info * _sta = NULL; \
+      _sta = (struct sta_info *)_param;\
+      \
+    }\
+    _reply = (clientStatusReply_t *)malloc(sizeof(*_reply) + _len); \
+             memset(_reply, 0, sizeof(*_reply)+ _len); \
+  } while(0); 
+
+
+typedef struct status_map_s
+{
+ char *status;
+ unsigned int val;
+}status_map_t;
+
+static int wpa_pac_send_data (char *buf, int bufLen) 
+{
+  struct sockaddr_in saddr;
+  int fd, ret_val;
+  struct hostent *local_host;	/* need netdb.h for this */
+  int i;
+  char *ptr = buf;
+  struct linger sl;
+  struct sockaddr_in client;
+  socklen_t clientlen = sizeof(client);
+
+  printf("buffer: ");  
+  for (i =0; i<10; i++)
+  {
+    printf("0x%x ", ptr[i]);  
+  }
+  printf("\n");  
+
+	/* Step1: create a TCP socket */
+	fd = socket(AF_INET, SOCK_STREAM, 0); 
+	if (fd == -1)
+	{
+		fprintf(stderr, "socket failed [%s]\n", strerror(errno));
+		return -1;
+	}
+	printf("Created a socket with fd: %d\n", fd);
+
+	/* Let us initialize the server address structure */
+	saddr.sin_family = AF_INET;         
+	saddr.sin_port = htons(3434);     
+	local_host = gethostbyname("127.0.0.1");
```

Comment:
I'm guessing this local socket is for the communication between "hostapd" & "pacd" applications, 

My question is why have not leveraged/added logic to  the control interface that is exposed by hostapd via lib_wpaclient 

https://gist.github.com/sepfy/25c1a34e9b418de230a1d97c56de88a6

Reply:
The control interface uses synchronous communication. The PAC uses the control interface to send command updates etc. In case of asynchronous data reception such as information retrieved from the radius response etc , PAC listens on a server socket. Hence this change.